src/ParseServer.js

```diff
@@ -64,73 +64,80 @@ class ParseServer {
-    databaseController
-      .performInitialization()
-      .then(() => hooksController.load())
-      .then(async () => {
-        const startupPromises = [];
-        if (schema) {
-          startupPromises.push(new DefinedSchemas(schema, this.config).execute());
-        }
-        if (
-          options.cacheAdapter &&
-          options.cacheAdapter.connect &&
-          typeof options.cacheAdapter.connect === 'function'
-        ) {
-          startupPromises.push(options.cacheAdapter.connect());
-        }
-        startupPromises.push(liveQueryController.connect());
-        await Promise.all(startupPromises);
-        if (serverStartComplete) {
-          serverStartComplete();
+  async start() {
+    try {
+      if (this.config.state === 'ok') {
+        return this;
+      }
+      this.config.state = 'starting';
+      Config.put(this.config);
+      const {
+        databaseController,
+        hooksController,
+        cloud,
+        security,
+        schema,
+        cacheAdapter,
+        liveQueryController,
+      } = this.config;
+      try {
+        await databaseController.performInitialization();
+      } catch (e) {
+        if (e.code !== Parse.Error.DUPLICATE_VALUE) {
+          throw e;
         }
-      })
-      .catch(error => {
-        if (serverStartComplete) {
-          serverStartComplete(error);
+      }
+      await hooksController.load();
+      const startupPromises = [];
+      if (schema) {
+        startupPromises.push(new DefinedSchemas(schema, this.config).execute());
+      }
+      if (cacheAdapter?.connect && typeof cacheAdapter.connect === 'function') {
+        startupPromises.push(cacheAdapter.connect());
+      }
+      startupPromises.push(liveQueryController.connect());
+      await Promise.all(startupPromises);
+      if (cloud) {
+        addParseCloud();
+        if (typeof cloud === 'function') {
+          await Promise.resolve(cloud(Parse));
+        } else if (typeof cloud === 'string') {
+          if (process.env.npm_package_type === 'module') {
+            await import(path.resolve(process.cwd(), cloud));
+          } else {
+            require(path.resolve(process.cwd(), cloud));
+          }
         } else {
-          console.error(error);
-          process.exit(1);
+          throw "argument 'cloud' must either be a string or a function";
         }
-      });
-
-    if (cloud) {
-      addParseCloud();
-      if (typeof cloud === 'function') {
-        cloud(Parse);
-      } else if (typeof cloud === 'string') {
-        require(path.resolve(process.cwd(), cloud));
-      } else {
-        throw "argument 'cloud' must either be a string or a function";
       }
-    }
-
-    if (security && security.enableCheck && security.enableCheckLog) {
-      new CheckRunner(options.security).run();
+      if (security && security.enableCheck && security.enableCheckLog) {
+        new CheckRunner(security).run();
+      }
+      this.config.state = 'ok';
```

Comment:
Is the config a good place to put the parse server state? Any reasons for this or alternatives?

Reply:
Yes, `config` is used to store all current state and settings related to Parse Server instances, so I think it's the best place

Reply:
Got it